clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1305,10 +1305,16 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
      */
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
-        return partition != null ?
-                partition :
-                partitioner.partition(
-                        record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        if (partition != null) {
+            return partition;
+        }
+
+        int customPartition = partitioner.partition(
+                record.topic(), record.key(), serializedKey, record.value(), serializedValue, cluster);
+        if (customPartition < 0) {
+            throw new IllegalArgumentException(String.format("Invalid partition: %d. Partition number should always be non-negative.", customPartition));
```

Comment:
Could you make the error message to be more specific? E.g. "Partitioner generated invalid partition: %d.." to differentiate that a partitioner is used.

Reply:
Sure. I have pushed a new commit. How about this one?
```java
"The partitioner generated an invalid partition number: %d. Partition number should always be non-negative."
```

By the way, I think Jenkins PR check did not run quite well. The failed tests are not related to this change. I tried some of these. Looks like they could be passed on my laptop.